rust/src/conflict_checker.rs

```diff
@@ -0,0 +1,459 @@
+//! Helper module to check if a transaction can be committed in case of conflicting commits.
+use crate::action::{Action, Add, DeltaOperation, MetaData, Protocol, Remove};
+use crate::{
+    table_state::DeltaTableState, DeltaDataTypeTimestamp, DeltaDataTypeVersion, DeltaTable,
+    DeltaTableError, DeltaTableMetaData,
+};
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+use std::collections::{HashMap, HashSet};
+use std::io::{BufRead, BufReader, Cursor};
+
+/// The commitInfo is a fairly flexible action within the delta specification, where arbitrary data can be stored.
+/// However the reference implementation as well as delta-rs store useful information that may for instance
+/// allow us to be more permissive in commit conflict resolution.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct CommitInfo {
+    pub version: Option<DeltaDataTypeVersion>,
+    pub timestamp: DeltaDataTypeTimestamp,
+    pub user_id: Option<String>,
+    pub user_name: Option<String>,
+    pub operation: String,
+    pub operation_parameters: HashMap<String, String>,
+    pub read_version: Option<i64>,
+    pub isolation_level: Option<IsolationLevel>,
+    pub is_blind_append: Option<bool>,
+}
+
+/// Exceptions raised during commit conflict resolution
+#[derive(thiserror::Error, Debug)]
+pub enum CommitConflictError {
+    /// This exception occurs when a concurrent operation adds files in the same partition
+    /// (or anywhere in an unpartitioned table) that your operation reads. The file additions
+    /// can be caused by INSERT, DELETE, UPDATE, or MERGE operations.
```

Comment:
I don't think `INSERT` / `INSERT` can result in this error. As far as I can understand from https://docs.databricks.com/optimizations/isolation-level.html#write-conflicts-on-databricks.

Maybe it will be useful to create that matrix from the given link and add it somewhere in this PR maybe on the enum explanation.